File: arrays/spiralTraverse/spiralTraverse.js
function spiralTraverse(array) {
  let startCol = 0, endCol =  array[0].length - 1
  let startRow = 0, endRow =  array.length - 1
  
  const spiralArray = []
  while(startCol <= endCol && startRow <= endRow) {
    
    for(let i = startCol; i <= endCol; i++ ) {
       spiralArray.push(array[startRow][i])
    }

    for(let i = startRow + 1; i <= endRow; i++){
      spiralArray.push(array[i][endCol])
    }

    if(endRow > startRow) {
      for(let  i = endCol - 1; i >= startCol; i--) {
          spiralArray.push(array[endRow][i])
      }
    }

    if(endCol > startCol) {
        for(let i = endRow - 1; i >= startRow + 1; i-- ) {
          spiralArray.push(array[i][startCol])
        }
    }

    startCol++
    endCol--
    startRow++
    endRow--
  }
    
return spiralArray
}
// O(n) time | O(n) space
console.log(spiralTraverse([
    [1, 3, 2, 5, 4, 7, 6]
  ]))

File: arrays/spiralTraverse/spiralTraverse.py
def spiralTraverse(array):
    result = []
    startRow, endRow = 0, len(array) - 1
    startCol, endCol = 0, len(array[0]) - 1

    while startRow <= endRow and startCol <= endCol:
        for col in range(startCol, endCol + 1):
            result.append(array[startRow][col])

        for row in range(startRow + 1, endRow + 1):
            result.append(array[row][endCol])

        if startRow < endRow:
            for col in reversed(range(startCol, endCol)):
                result.append(array[endRow][col])
        if startCol < endCol:
            for row in reversed(range(startRow + 1, endRow)):
                result.append(array[row][startCol])

        startRow += 1
        endRow -= 1
        startCol += 1
        endCol -= 1

    return result


# O(n) time | O(n) space

print(spiralTraverse([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12], [13, 14, 15, 16]]))
# Expected output: [1, 2, 3, 4, 8, 12, 16, 15, 14, 13, 9, 5, 6, 7, 11, 10]
